java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java

```diff
@@ -17,81 +17,242 @@
-  }
+    private boolean allocated = false;
+
+    /**
+     * Note: fields are not closed when the writer is closed
+     *
+     * @param schema
+     * @param out
+     * @param allocator
+     */
+    protected ArrowWriter(Schema schema, OutputStream out, BufferAllocator allocator) {
+        this(schema.getFields(), createVectors(schema.getFields(), allocator), Channels.newChannel(out), true);
+    }
+
+    protected ArrowWriter(Schema schema, WritableByteChannel out, BufferAllocator allocator) {
+        this(schema.getFields(), createVectors(schema.getFields(), allocator), out, true);
+    }
+
+    protected ArrowWriter(List<Field> fields, List<FieldVector> vectors, OutputStream out) {
+        this(fields, vectors, Channels.newChannel(out), false);
+    }
+
+    protected ArrowWriter(List<Field> fields, List<FieldVector> vectors, WritableByteChannel out, boolean allocated) {
+        this.vectors = vectors;
+        this.out = new WriteChannel(out);
+        this.allocated = allocated;
+
+        // translate dictionary fields from in-memory format to message format
+        // add dictionary ids, change field types to dictionary type instead of index type
+        List<Field> updatedFields = new ArrayList<>(fields);
+        List<ArrowDictionaryBatch> dictionaryBatches = new ArrayList<>();
+        Set<Long> dictionaryIds = new HashSet<>();
+
+        // go through to add dictionary id to the schema fields and to unload the dictionary batches
+        for (FieldVector vector: vectors) {
+            if (vector instanceof DictionaryVector) {
+                Dictionary dictionary = ((DictionaryVector) vector).getDictionary();
+                long dictionaryId = dictionary.getId();
+                Field field = vector.getField();
+                // find the dictionary field in the schema
+                Field schemaField = null;
+                int fieldIndex = 0;
+                while (fieldIndex < fields.size()) {
+                    Field toCheck = fields.get(fieldIndex);
+                    if (field.getName().equals(toCheck.getName())) { // TODO more robust comparison?
+                        schemaField = toCheck;
+                        break;
+                    }
+                    fieldIndex++;
+                }
+                if (schemaField == null) {
+                    throw new IllegalArgumentException("Dictionary field " + field + " not found in schema " + fields);
+                }
+
+                // update the schema field with the dictionary type and the dictionary id for the message format
+                ArrowType dictionaryType = dictionary.getVector().getField().getType();
+                Field replacement = new Field(field.getName(), field.isNullable(), dictionaryType, dictionary.getEncoding(), field.getChildren());
+
+                updatedFields.remove(fieldIndex);
+                updatedFields.add(fieldIndex, replacement);
```

Comment:
Let me clarifies this.
We should not add a type to the Messages.fbs for dictionary. I think that was part of the original PR and was rolled back.
The MinorType class is more of a Vector Class ID. It should be renamed IMO but this is a separate discussion. In my comment I said "possibly" meaning I was not quite sure, so let's take the MinorType comment out for now. 
Here is what I was thinking about:
We need to change the implementation of FieldVector.initializeChildrenFromFields() [1]. Look in particular at the implementation in MapVector [2] that uses AbstractMapVector.add [3] which uses MinorType.getNewVector [4] to create a new vector.
initializeChildrenFromFields is responsible instantiating the right Vector implementation based on each Field in the Schema. So really [2]/[3] should be changed to create a DictionaryVector instead of the type that would be used if not dictionary encoded.
That would replace some of the logic in ArrowReader.initialize() in this PR. This may not require a new MinorType, you'll see when you're in there.

[1] https://github.com/apache/arrow/blob/d28f1c1e0f21bc578b84ab4bed4cf259c333fbc9/java/vector/src/main/java/org/apache/arrow/vector/FieldVector.java#L38
[2] https://github.com/apache/arrow/blob/d28f1c1e0f21bc578b84ab4bed4cf259c333fbc9/java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java#L314
[3] https://github.com/apache/arrow/blob/d28f1c1e0f21bc578b84ab4bed4cf259c333fbc9/java/vector/src/main/java/org/apache/arrow/vector/complex/AbstractMapVector.java#L170
[4] https://github.com/apache/arrow/blob/d28f1c1e0f21bc578b84ab4bed4cf259c333fbc9/java/vector/src/main/java/org/apache/arrow/vector/types/Types.java#L564